gtsam/navigation/tests/testNavStateImuEKF.cpp

```diff
@@ -0,0 +1,276 @@
+  CHECK(trace_pos_post < trace_pos_prior);
+
+  // Compare EKF position update to solving an equivalent GaussianFactorGraph.
+  using symbol_shorthand::X;
+  const Key key = X(0);
+  GaussianFactorGraph gfg;
+
+  // Prior as JacobianFactor using full Gaussian covariance:
+  // || I * x - 0||_{P^{-1}}
+  Vector b = Vector::Zero(9);
+  gfg.add(key, I_9x9, b, noiseModel::Diagonal::Variances(P_prior.diagonal()));
+
+  // Measurement as JacobianFactor using full Gaussian covariance:
+  // || H * x -innovation ||_{R^{-1}}
+  Vector b_meas = innovation;  // 3x1
+  gfg.add(key, H, b_meas, noiseModel::Isotropic::Sigma(3, sigma));
+
+  // Solve for MAP delta
+  VectorValues delta_map = gfg.optimize();
+  const Vector delta_graph = delta_map.at(key);
+
+  // THEN: Graph solution equals EKF correction
+  EXPECT(assert_equal(delta_graph, delta_ekf, 1e-9));
+}
+
+/* ************************************************************************* */
+NavState oldNavStateImuDynamics(const NavState& X, const Vector3& omega_b,
+                                const Vector3& f_b, double dt,
+                                const Vector3& n_gravity,
+                                OptionalJacobian<9, 9> H = {}) {
+  // Rotation and velocity
+  const Rot3& R = X.attitude();
+  Matrix3 D_vb_R, D_vb_v, D_gb_R;
+  const Vector3& v_n = X.velocity();  // Has D_v_v = R !
+  const Vector3 v_body =
+      R.unrotate(v_n, H ? &D_vb_R : nullptr, H ? &D_vb_v : nullptr);
+  const Vector3 g_body = R.unrotate(n_gravity, H ? &D_gb_R : nullptr);
+  const Vector3 a_b_total = f_b + g_body;
+
+  // Construct increment directly as group element with body-frame p/v
+  // increments
+  const Rot3 dR = Rot3::Expmap(omega_b * dt);
+  const double dt2 = 0.5 * dt * dt;
+  const Vector3 dp_body = v_body * dt + a_b_total * dt2;
+  const Vector3 dv_body = a_b_total * dt;
+  NavState U(dR, dp_body, dv_body);
+
+  if (H) {
+    Matrix3 dRt = dR.transpose();  // Jacobian of NavState::Create
+    H->setZero();
+    // position:
+    H->template block<3, 3>(3, 0) = dRt * (D_vb_R * dt + D_gb_R * dt2);
+    const Matrix3 D_v_v = R.matrix();  // Jacobian of velocity()
+    H->template block<3, 3>(3, 6) = dRt * D_vb_v * D_v_v * dt;
+    // velocity:
+    H->template block<3, 3>(6, 0) = dRt * D_gb_R * dt;
+  }
+
+  return U;
+}
+
+TEST(NavStateImuEKF, WPsiU_matches_navStateImuDynamics) {
+  using namespace nontrivial_navstate_example;
+
+  const double dt = 1e-2;  // 10 ms
+  const Vector3 n_gravity(0, 0, -9.81);
+
+  // Expected state from oldNavStateImuDynamics
+  Matrix9 D_U_X0, D_X_X0, D_X_U;
+  NavState U = oldNavStateImuDynamics(X0, omega_b, f_b, dt, n_gravity, D_U_X0);
+  NavState X_old = X0.compose(U, D_X_X0, D_X_U);
+
+  // Check New dynamics function
+  Matrix9 A_ekf;
+  NavState X_ekf =
+      NavStateImuEKF::Dynamics(n_gravity, X0, omega_b, f_b, dt, A_ekf);
+
+  CHECK(assert_equal(X_old, X_ekf));
+  CHECK(assert_equal<Matrix>(D_X_X0 + D_X_U * D_U_X0, A_ekf));
```

Comment:
This line proves a hunch I had, that the old LieGroupEKF version, which is tested here, canceled the gravity pieces of the Jacobian A. PS D_X_X0 is Adjoint(inv(U)) and D_X_U = identity.